Post array events against the field pointer, not the array pointer

krisztianloki opened the pull request:
This fixes github issue #97:
Reading into an aai record from a compress or histogram or subArray record
could cause a segfault if the aai record was initialized before the
linked one.

krisztianloki committed:
Allow changing the BPTR field

This fixes github issue #97:
Reading into an aai record from a compress or histogram or subArray record
could cause a segfault if the aai record was initialized before the
linked one.

mdavidsaver commented:
This all looks good.  (I see that I missed subArray yesterday)  I do notice that that there is a second `db_post_events()` in histogram which needs changing.

```
$ git grep db_post_events.*bptr
modules/database/src/std/rec/compressRecord.c:    db_post_events(prec, prec->bptr, monitor_mask);
modules/database/src/std/rec/histogramRecord.c:        db_post_events(prec, prec->bptr, DBE_VALUE | DBE_LOG);
modules/database/src/std/rec/histogramRecord.c:        db_post_events(prec, prec->bptr, monitor_mask);
modules/database/src/std/rec/subArrayRecord.c:    db_post_events(prec, prec->bptr, monitor_mask);
```

krisztianloki committed:
Fixed db_post_events to not use bptr

anjohnson commented:
Since the bug that this PR fixes could still exist in external array record types too, should we add a check and warning to the `dbEntryToAddr()` routine in in dbAccess.c if a record's `cvt_dbaddr()` method changes `paddr->pfield`? That's the only place this method is called and it can easily detect the cases we're concerned about. I don't think this change would break backwards-compatibility with older Base releases:
```patch
--- a/modules/database/src/ioc/db/dbAccess.c
+++ b/modules/database/src/ioc/db/dbAccess.c
@@ -621,7 +621,13 @@ long dbEntryToAddr(const DBENTRY *pdbentry, DBADDR *paddr)
 
         /* Let record type modify paddr */
         if (prset && prset->cvt_dbaddr) {
-            return prset->cvt_dbaddr(paddr);
+            long status = prset->cvt_dbaddr(paddr);
+            if (paddr->pfield != pdbentry->pfield) {
+                errlogPrintf("Warning: %s::cvt_dbaddr('%s') set paddr->pfield\n"
+                    "  See https://github.com/epics-base/epics-base/pull/114\n",
+                    pdbentry->precordType->name, paddr->precord->name);
+            }
+            return status;
         }
     }
     return 0;
```
The message points to this PR which should make it obvious what changes are needed in the record code and already points back to the problem that this fixes. Should I limit how many times it appears? I may adjust the PR's title and/or description slightly to make it more obvious though.

@krisztianloki I'm not asking you to do any more with this PR, I'm in the process of merging it (by hand) and this idea struck me while I was doing that.

mdavidsaver commented:
> ... should we add a check and warning to the `dbEntryToAddr()` routine ...

I wondered about this as well.  I was thinking about a error check of `status || pfield!=NULL` after `cvt_dbaddr()` to detect this specific condition.  A stricter warning seems likely to produce a lot of noise in an IOC with eg. an `sscanRecord`.  I think it would be reasonable if it could be limited to say, once per field per record type (maybe by hiding a flag in `dbFldDes`?).

anjohnson commented:
The warning is only shown when something (either a DB link or a network client) actually connects to that field, but I like the idea of a flag in dbFldDes to make it only display once, so I've added that (ignoring the fact that there are no locks protecting that structure so _in theory_ this could race if RSRV and QSRV searches collide on it).

Merging.

mdavidsaver commented:
> Merging.

Maybe not yet?  This might be worth a courtesy warning to the synapps gang, since they will likely be the first to see the warnings, and receive the complaints.

anjohnson commented:
The potential for a segfault when an aai.INP points to any such field dates back several releases so I want to put the warning detection code in, but I will enable it only if `getenv("EPICS_PFIELD_WARNING")` exists. Initially I thought the synapps modules looked like a lot of work, but only 12 of the 25 record types actually have a `cvt_dbaddr()` method. We missed 4 of the built-in record types too – bptr isn't the only field used to hold array pointers...

Checking and fixing all the synapps record types that have that method would be an excellent codeathon task:
- allenBradley/allenBradleyApp/1771DCMSrc/abDcmRecord.c
- allenBradley/allenBradleyApp/1771IFESrc/ab1771IFERecord.c
- allenBradley/allenBradleyApp/1771IXSrc/ab1771IXRecord.c
- allenBradley/allenBradleyApp/1771NSeriesSrc/ab1771NRecord.c
- asyn/asyn/asynRecord/asynRecord.c
- calc/calcApp/src/aCalcoutRecord.c
- calc/calcApp/src/sCalcoutRecord.c
- camac/camacApp/src/camacRecord.c
- mca/mcaApp/mcaSrc/mcaRecord.c
- optics/opticsApp/src/tableRecord.c
- sscan/sscanApp/src/sscanRecord.c
- vme/vmeApp/src/vmeRecord.c

anjohnson commented:
Change of plan. There's a major problem with requiring the pfield change to fields that hold long strings (lsi, lso, printf). The way that these work is that they present as a DBF_STRING but with a field size matching the buffer length, and the `$` field modifier knows that it can convert them into an array of that length. The `cvt_dbaddr()` method thus _has_ to modify the `paddr->pfield` pointer to point to the buffer in case the the field is accessed as a `DBF_STRING`, in which case there is no subsequent call to `get_array_info()`.

The changes in this PR are still fine so I will probably merge them along with some slightly more complicated equivalent changes to the aSub record type, although I need sleep first.

We should try to get the synapps record types similarly fixed at the codeathon, but the check and warning aspects just aren't possible without doing more work, so later if at all.

anjohnson committed:
Modify aai to support pass-1 device initialization

The Soft Channel device support requests pass-1 initialization.
It no longer needs to initialize the INP link or allocate the
array buffer itself, these are taken care of elsewhere.

The record code uses PACT to remember that the device must be
initialized again in pass 1.

anjohnson committed:
Release notes and aai documentation updates

anjohnson commented:
I'm not happy with asking all of the external record types to be changed just because the aai record has some strange semantics. I have a solution that doesn't require that, although the changes in this PR are still sensible and I will include them.

The Soft Channel support's `init_record()` routine can now advise the record (by returning the status value 2) that it needs to be called again in pass 1. The record will make temporary use of the PACT field to remember that it needs to reinitialize the device support. The updated Soft Channel device support can handle its constant initialization duties in pass 1 without demanding special behavior from the other record types. Existing device support will continue to work as before since it has always returned 0, -1, or an error status value – 2 would have been treated as an error.

I have added a Release Note, and updated the record reference documentation in the .dbd.pod file to describe the new initialization semantics. I am pushing my changes to this branch now, and plan to merge them tomorrow if there are no major objections.

mdavidsaver commented:
I'm not thrilled with another magic `return 2;`, but agree that this is preferable to trying to force a change to the external record types.  As far as I can see, this wouldn't break existing device support using HW link types (I've written some).

> The record will make temporary use of the PACT field to remember that it needs to reinitialize the device support

No objection.  fyi. This kind of extra/private state could instead be added as a bit field in `struct dbCommonPvt`.

anjohnson committed:
Replace magic "2" with macro AAI_DEVINIT_PASS1

anjohnson commented:
> a bit field in `struct dbCommonPvt`

Since in this case it's specific to the aai record type, it would make more sense to add another field to the record type if I couldn't use PACT for it.

I have replaced the magic `2` with a macro `AAI_DEVINIT_PASS1` defined in the aaiRecord.h file.

Thanks.
